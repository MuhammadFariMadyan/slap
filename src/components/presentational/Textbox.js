import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import "typeface-roboto";

import ContentWrapper from "./ContentWrapper";

export const Text = styled.input`
  font-size: 32px;
  font-family: Roboto;
  color: #333;
  width: ${({ width }) => width}px;
  height: ${({ height }) => height}px;
  padding: 0px;
  border: 0px;
  padding: 8px;
  box-sizing: border-box;
  outline: none;
  background: white;
  cursor: ${({ isActive, disabled }) =>
    isActive && disabled ? "pointer" : "auto"};
`;

const Textbox = ({
  value,
  width,
  height,
  x,
  y,
  onClick,
  onDoubleClick,
  onChange,
  isActive,
  isEditable
}) => (
  <g onClick={onClick} onDoubleClick={onDoubleClick}>
    <foreignObject width={width} height={height} x={x} y={y}>
      <Text
        width={width}
        height={height}
        value={value}
        onChange={onChange}
        disabled={!isEditable}
        isActive={isActive}
      />
    </foreignObject>
    {isActive && <ContentWrapper width={width} height={height} x={x} y={y} />}
  </g>
);

Textbox.propTypes = {
  value: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
  x: PropTypes.number,
  y: PropTypes.number,
  onClick: PropTypes.func,
  onDoubleClick: PropTypes.func,
  onChange: PropTypes.func,
  isActive: PropTypes.bool,
  isEditable: PropTypes.bool
};

export default Textbox;
