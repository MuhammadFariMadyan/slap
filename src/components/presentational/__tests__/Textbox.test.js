import React from "react";
import { shallow } from "enzyme";
import "jest-styled-components";

import Textbox, { Text } from "../Textbox";

describe("Textbox", () => {
  const setup = propOverrides => {
    const props = {
      value: "My Text",
      width: 420,
      height: 48,
      x: 24,
      y: 24,
      onChange: jest.fn(),
      onClick: jest.fn(),
      onDoubleClick: jest.fn(),
      ...propOverrides
    };

    const wrapper = shallow(<Textbox {...props} />);

    return {
      wrapper,
      props
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Text)).toHaveStyleRule("cursor", "auto");
  });

  it("should renders in active", () => {
    const { wrapper } = setup({ isActive: true });

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Text)).toHaveStyleRule("cursor", "pointer");
  });

  it("should renders in editable", () => {
    const { wrapper } = setup({ isActive: true, isEditable: true });

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Text)).toHaveStyleRule("cursor", "auto");
  });

  it("should calls 'onClick' callback while clicked", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("click");

    expect(props.onClick).toBeCalled();
  });

  it("should calls 'onDoubleClick' callback while double clicked", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("dblclick");

    expect(props.onDoubleClick).toBeCalled();
  });

  it("should calls 'onChange' callback while changed", () => {
    const { wrapper, props } = setup({ isActive: true, isEditable: true });
    const event = {
      target: {
        value: "Testing"
      }
    };

    wrapper.find(Text).simulate("change", event);

    expect(props.onChange).toBeCalledWith(event);
  });
});
