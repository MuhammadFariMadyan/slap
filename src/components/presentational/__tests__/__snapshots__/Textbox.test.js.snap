// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textbox should renders default 1`] = `
<g
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <foreignObject
    height={48}
    width={420}
    x={24}
    y={24}
  >
    <styled.input
      disabled={true}
      height={48}
      onChange={[Function]}
      value="My Text"
      width={420}
    />
  </foreignObject>
</g>
`;

exports[`Textbox should renders in active 1`] = `
<g
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <foreignObject
    height={48}
    width={420}
    x={24}
    y={24}
  >
    <styled.input
      disabled={true}
      height={48}
      isActive={true}
      onChange={[Function]}
      value="My Text"
      width={420}
    />
  </foreignObject>
  <ContentWrapper
    height={48}
    width={420}
    x={24}
    y={24}
  />
</g>
`;

exports[`Textbox should renders in editable 1`] = `
<g
  onClick={[Function]}
  onDoubleClick={[Function]}
>
  <foreignObject
    height={48}
    width={420}
    x={24}
    y={24}
  >
    <styled.input
      disabled={false}
      height={48}
      isActive={true}
      onChange={[Function]}
      value="My Text"
      width={420}
    />
  </foreignObject>
  <ContentWrapper
    height={48}
    width={420}
    x={24}
    y={24}
  />
</g>
`;
